### ISSUE TITLE
Questions regarding generate_adv_images.py

### ISSUE THREAD

**Issue by TalRub104** [HashmatShadab/MambaRobustness#1]
Hi,
I have two questions regarding generate_adv_images.py:
1. Is your 'Normalize' class suitable for the CIFAR-10 dataset as well?

2. It appears that your 'Normalize' class doesn't guarantee the pixel range to be [0, 1], as required by the attack. In line 22 of pgd.py:
images: :math:(N, C, H, W) where N is the number of batches, C is the number of channels, H is the height, and W is the width. The image data must have a range of [0, 1].

Can you explain why your code works given you Normalize class?

3. Can you share how long it took you to apply the attack on the entire ImageNet validation set? Have you considered applying the attack to only 1,000 images to save time?

**Comment by HashmatShadab** [comment]
Hi @TalRub104 



1. The `Normalize` class is used for input image normalization, following the same mean and standard deviation values that were used to normalize images during the pretrained model’s training. Same can be done for CIFAR-10, you just have to pass the mean and standard deviation for the pretrained model.

2. The `Normalize` class is embedded within the model rather than the data preprocessing pipeline to enable adversarial perturbation generation on the input image, which is in the [0, 1] range. By integrating Normalize into the model, we can backpropagate through it, allowing for gradient computation on the input images(which is in range [0, 1]).

3. Adversarial images were generated using a subset of 5,000 images from the ImageNet validation set.

**Comment by TalRub104** [comment]
Thanks for the quick response.
I am working with the Vim model (https://github.com/hustvl/Vim)
and there they do not use an embedded Normalize class but rather 
use the following data preprocessing (datasets.py):
```
def build_transform(is_train, args):
    resize_im = args.input_size > 32
    if is_train:
        # this should always dispatch to transforms_imagenet_train
        transform = create_transform(
            input_size=args.input_size,
            is_training=True,
            color_jitter=args.color_jitter,
            auto_augment=args.aa,
            interpolation=args.train_interpolation,
            re_prob=args.reprob,
            re_mode=args.remode,
            re_count=args.recount,
        )
        if not resize_im:
            # replace RandomResizedCropAndInterpolation with
            # RandomCrop
            transform.transforms[0] = transforms.RandomCrop(
                args.input_size, padding=4)
        return transform

    t = []
    if resize_im:
        size = int(args.input_size / args.eval_crop_ratio)
        t.append(
            transforms.Resize(size, interpolation=3),  # to maintain same ratio w.r.t. 224 images
        )
        t.append(transforms.CenterCrop(args.input_size))

    t.append(transforms.ToTensor())
    t.append(transforms.Normalize(IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD)) 
    return transforms.Compose(t)
```

If I trained a Vim model on ImageNet/CIFAR-10 using the above transform and now want to check the model's validation accuracy on adversarial images (using PGD, for example), would it be okay to use your script (including the embedded Normalize class approach), or do I need to have used your approach during training as well? I'm asking because I have already completed the training and I’m wondering if I can just use your script to evaluate adversarial robustness.


Regarding your claim that the Normalize class will make the input to the model fall within the range [0, 1], could you explain this? For example, if I have a pixel value of 0.3 and perform the operation (0.3 - 0.485) / 0.229, I get a value of -0.8.


Thanks!

**Comment by HashmatShadab** [comment]
1. The `Normalize` class is included here to enable backpropagation through the normalization process, as we need to compute gradients with respect to the input image for generating adversarial attacks. This technique is commonly employed in adversarial attack methods. Similar to training, the model processes a normalized version of the input image, but instead of normalizing during data preprocessing, we apply it directly within the model. For standard training or evaluation, there's no need to include the `Normalize` class inside the model, as normalization is handled in the data preprocessing pipeline, and computing gradients with respect to the input image isn't necessary.

2. I'm not suggesting that the input after normalization will be in the range [0,1]. The input is in the range [0,1] before normalization, and the gradients are calculated based on this pre-normalized input.

You can check this easily by debugging the code and checking the input image on which the perturbations are added.

**Comment by TalRub104** [comment]
And why do you use the Normalize class for the imagenet-a inference? why not just using:
test_transform = transforms.Compose([transforms.Resize(256), transforms.CenterCrop(224), transforms.ToTensor(), transforms.Normalize(mean, std)]) as performed in the original imagenet-a github "https://github.com/hendrycks/natural-adv-examples" ?

In additon, is there any special reason you created pgd.py, fgsm.py instead of using "from torchattacks import PGD, FGSM" ? Your code looks the same to the one in torchattacks.

**Comment by HashmatShadab** [comment]
1. Applying normalization either during data preprocessing or as the initial step before passing it to the model does not affect the outcome. You can use either approach for standard inference.  
2. Yes, the attack code comes from Torchattacks and has been integrated into the codebase for convenience, allowing modifications if needed.
